### ISSUE TITLE
Need ways to check whether segment reload/refresh is done on server side

### ISSUE THREAD

**Issue by jackjlli** [apache/pinot#8492]
Currently segment reload/refresh is done asynchronously. There is no way for Pinot admin to know whether the segment reload/refresh request is fully done on server side. It could be possible that multiple reload/refresh are made, which can cause race conditions.

We need some way to figure out the state of the reload/refresh request on server side.

Some thoughts: 
* Add server level APIs to fetch the state.
* Add callback when the request is done.

**Comment by Jackie-Jiang** [comment]
Since the reload/refresh is done via the Helix messaging mechanism, we should also explore whether Helix can support reading response for the message.

**Comment by mcvsubbu** [comment]
I thought we already added several server level APIs to fetch the state exactly for this purpose? Are you sure it is not there already?

**Comment by mcvsubbu** [comment]
Can we use this API? https://github.com/apache/pinot/blob/master/pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java

**Comment by Jackie-Jiang** [comment]
@mcvsubbu We built this API as a work-around to check the reload status, but it does have some limitations/drawbacks:
- Not able to show refresh status
- Not user friendly. Users need to know which index to check in order to know if the segment is reloaded. They also need some client side code to automate the check.

IMO it would be good to add APIs to directly tell how many segments out of the total segments are reloaded/refreshed, or at least a flag tracking whether it is done.

**Comment by jackjlli** [comment]
It's not very intuitive nor straightforward to retrieve other information in order to check the state of a reload/refresh request. 

What's worse, one corner scenario I can think of is that if table config remains unchanged while a reload request is sent, the segment metadata remains unchanged, but there could be that reload request is still in progress. In this case, we cannot determine whether the reload request is done by just validating metadata of a segment. And this actually happens in our existing integration test.

**Comment by jackjlli** [comment]
We do have a `SegmentRefreshSemaphore` class which is initialized only once per pinot-server, something that we can use to check the number of segment reloads / refreshes per server.

**Comment by Jackie-Jiang** [comment]
> We do have a `SegmentRefreshSemaphore` class which is initialized only once per pinot-server, something that we can use to check the number of segment reloads / refreshes per server.

This semaphore is used to control the parallelism of the reload/refresh, but not the segments already reloaded/refreshed. We need a different mechanism to return that info, e.g. associate each reload/refresh with an id and send the id within the message

**Comment by mcvsubbu** [comment]
> 

Yes, @jackjlli and I had a discussion yesterday, and reached the same conclusion. Using some sort of an ID. The problem is that the threads taking up any of these activities can be in any order. [Of course, we may choose not to handle multiple simultaneous requests]. In this case, it is useful to know the _highest_ sequence number of messages handled. 

And then we need to consider restarts. If a server restarts, then we should treat it as if all messages are handled. Then we may enter into race conditions when a message may never be received (or discarded after receipt), etc. Not very straight-forward.

A more complete solution will be to keep a mirror of the IDEALSTATE znode in zk (or, if Helix 1.x supports additional data along with idealstate). For example,
`"segment_0": { "host1": {"ONLINE", "105"}, "host2":{"ONLINE", 106"}}`
In this case, it indicates that host 1 processed message 105 whereas host2 processed 106.
Unless helix 1.x supports an idealstate extension natively, this can lead to a new znode, something we have tried to avoid in the past.

I suggest that we NOT implement this API. Instead, we consider the use cases.

Most common cases:
- operators may issue a single reload, and can use the segment status API to get the status
- operators almost never will try to get statuses and verify that they are the same if no reload is issued. So, we either remove this from the test code, or just put a sleep. But let us not code for this to work right.
- If operators indeed issue multiple reloads for the (segments of) same table, the code will work fine as long as the bug that has been found is fixed (the bug where we load the table config before grabbing the semaphore).

Just fix the bug, and document it well (that if users have 1000 servers for a single table, and do a reload, the zookeeper server may be overloaded with 1000 requests coming in at the same time.)

**Comment by npawar** [comment]
FYI @saurabhd336
